client/landing/stepper/declarative-flow/internals/hooks/use-step-navigation-with-tracking/index.ts

```diff
@@ -12,33 +14,35 @@ interface Params< FlowSteps extends StepperStep[] > {
 }
 
 export const useStepNavigationWithTracking = ( {
-	flow,
+	flow: { useTracksEventProps, useStepNavigation, name: flowName, variantSlug },
```

Comment:
Oh man, destructuring this broke the function binding. Some flows depend on `this` in `useStepNavigation` and connect domain is one of them.

So we either keep them structured or rebind the functions. You can visit `/setup/connect-domain?domain=[foo]&provider=[bar]` and see the issue.

Reply:
agh - I should have known. no excuse. put it back / no destructuring

I wish there was a way to avoid this situation. I do feel an inch scared to use hooks that a rules engine / editor cannot flag in the typical way. 
Was there any other proposal or experiment with not attaching things to `flow` object?

Reply:
Not really. I'm not sure what a better alternative would be.

Reply:
I don't know how much more of this we do at the framework level, but even the fact that ESList could not flag the violation feels pretty severe. The rest stems from there. The workaround leads to the `this` situation and there's no way around it. Unfortunately, these issues could have slipped into production without noticing/flagging.

Here's what I see:

- we've had a hook violation [here](https://github.com/Automattic/wp-calypso/pull/93273#discussion_r1709571056) that was ignored
- destructuring the `flow` param leads to `this` being `undefined` as there's usage of it within hooks to query the enclosing scope. 
  - This isn't normal. By convention, Hooks should be pure functions. Any usage outside of this norm increases complexity and falls against best practices. We have a series of hooks here bound to an object and assuming its context. I can even modify `this.name` or even `this.useSteps` and everything will "seem" to work until it doesn't.
- we use hooks dynamically by passing them around, which is akin to dependency injection. There's mention about it in official docs (as something to avoid) - it's no different when you pass the enclosing object around (worse I think): https://react.dev/reference/rules/react-calls-components-and-hooks#dont-dynamically-use-hooks

> what a better alternative would be

At the very least seek to align better with React best practices at the core of this framework. This will helps in long-term maintenance and avoid unexpected bugs from cropping up (we've seen two of them in this simple extension already).

`useStepNavigation( 'flow slug' )`, `useSteps( 'flow slug' )`, etc. come to mind.

> People are fair to assume...

Yes. I don't see how this "assumption" is reflected in the code though. The only bit of code that takes the assumption to heart is https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/site-setup-wg-flow.ts, which also mutates other structures in ways that look suspicious actually